src/messages/MOSDPing.h

```diff
@@ -81,40 +79,49 @@ class MOSDPing : public Message {
     bufferlist::iterator p = payload.begin();
     ::decode(fsid, p);
     ::decode(map_epoch, p);
-    ::decode(peer_as_of_epoch, p);
-    ::decode(op, p);
-    ::decode(peer_stat, p);
+    if (header.version < 3) {
+      osd_peer_stat_t peer_stat;
+      epoch_t peer_as_of_epoch;
+      ::decode(peer_as_of_epoch, p);
+      ::decode(op, p);
+      ::decode(peer_stat, p);
+    } else {
+      ::decode(op, p);
+    }
     ::decode(stamp, p);
     if (header.version >= 3) {
-      bufferlist size_bl;
       int payload_mid_length = p.get_off();
-      ::decode(size_bl, p);
-      min_message_size = size_bl.length() + payload_mid_length;
+      uint32_t size;
+      ::decode(size, p);
+      p.advance(size);
+      min_message_size = size + payload_mid_length;
     }
   }
   void encode_payload(uint64_t features) override {
     ::encode(fsid, payload);
     ::encode(map_epoch, payload);
-    ::encode(peer_as_of_epoch, payload);
     ::encode(op, payload);
-    ::encode(peer_stat, payload);
```

Comment:
You can't reorder these without making it a backwards-incompatible change ( COMPAT_VERSION).